Kubernetes v1.37 DRA feature blog (#56351)
kubernetes/website: content/en/blog/_posts/2026/kubernetes-v1-37-dra-updates.md

Reviewer: Nit for the future: please line-wrap instead of using one line per paragraph. Don't reflow now, it would make all pending comments disappear!

Reviewer 2: Per the Kubernetes style guide, we now write Alpha, Beta, Stable in title case - see https://kubernetes.io/docs/contribute/style/style-guide/#use-start-case-for-enhancement-graduation-phases

@KashishV999 I know some folks asked for all-lowercase but could you change it all back?

I'll dive into what's new for DRA in Kubernetes 1.37!

## What's stable in 1.37

[DRA Extended Resource support](https://www.kubernetes.dev/resources/keps/5004) has graduated to GA. This is the mechanism that lets DRA drivers satisfy requests made through the traditional extended resource API, think `example.com/gpu` in a Pod spec, without requiring a separate device plugin alongside the DRA driver. An extended resource name can be set directly on a DeviceClass, and Pods requesting it get matched to a device through DRA with no ResourceClaim needed on the workload's part.

It's been on a steady path since KEP acceptance in 1.34. alpha landed in 1.35, beta in 1.36, and now it's stable. For cluster operators, this is what makes DRA adoption gradual. Existing workloads written against extended resources keep working unmodified while the backend allocation logic moves over to DRA.

> [Resource Claim Status with possible standardized network interface data](https://www.kubernetes.dev/resources/keps/4817/) adds a `Devices` field to `ResourceClaim.Status`, letting DRA drivers report per-device status, including, for network devices, the interface name, MAC address, and IP addresses. This gives users and controllers visibility into device state that was previously invisible once a device was configured in a Pod, and makes it possible to build things like network services that rely on a device's reported IPs.

Reviewer 2: ```suggestion
[ResourceClaim status with possible standardized network interface data](https://www.kubernetes.dev/resources/keps/4817/) adds a `devices` field to ResourceClaim `.status`, letting DRA drivers report per-device status, including, for network devices, the interface name, MAC address, and IP addresses. This gives users and controllers visibility into device state that was previously invisible once a device was configured in a Pod, and makes it possible to build things like network services that rely on a device's reported IPs.
```

> [DRA: device taints and tolerations](https://www.kubernetes.dev/resources/keps/5055/) is now stable, DRA drivers can  mark devices as tainted so they're skipped for new Pod scheduling, and cluster admins can apply the same taints cluster-wide via a `DeviceTaintRule`, without reconfiguring drivers. Pods already using a tainted device can be evicted automatically, unless their `ResourceClaim` explicitly tolerates the taint. This mirrors node taints and tolerations, letting operators take a single device offline for maintenance or mark it degraded, without disrupting the rest of the cluster.

Reviewer 2: (There should be) no backticks around DeviceTaintRule.

custom rules. Before this, pairing devices from different vendors (like a GPU/TPU and a NIC on the same NUMA 
node) only worked if both drivers used the exact same attribute name. If one used `numa` and the other used 
`numaNode`, the scheduler couldn't pair them together. Now, you can easily bridge these differences yourself 
inside your manifest, meaning you don't have to wait for hardware vendors to agree on standardized attribute 
names. Beyond just fixing naming differences, you can also use CEL to handle more complex scenarios like 
slicing a specific ID out of a long, monolithic topology string, or grouping devices into custom performance 
tiers based on their available capacity.

[DRA Device Compatibility Groups](https://www.kubernetes.dev/resources/keps/5963) lets drivers tag partitions 
of a device, like MIG vs vGPU profiles on the same GPU, with compatibility groups, so the scheduler rejects 
incompatible combinations up front instead of the driver failing at node preparation time. It's controlled by 
the `DRADeviceCompatibilityGroups` feature gate, disabled by default.

> [PreQueueingHint extension point](https://www.kubernetes.dev/resources/keps/6132) is new as alpha in 1.37. 

Reviewer 3: This is new, but beta. 

See https://github.com/kubernetes/kubernetes/pull/138916.

https://github.com/kubernetes/kubernetes/blob/e81f39c0e03ce8ed8e2660c9147b391edd9e262b/pkg/features/kube_features.go#L2122-L2124

Reviewer 3: And this should be moved to the beta section.

Reviewer: https://github.com/kubernetes/kubernetes/blob/e81f39c0e03ce8ed8e2660c9147b391edd9e262b/pkg/features/kube_features.go#L2122-L2124 shows it as alpha, which is correct and intended.

The original plan was beta, but it got demoted to alpha after some quality concerns were raised.

Reviewer 3: Got it. 

Let me update the enhancement issue there as https://github.com/kubernetes/kubernetes/pull/140959 .

DRA ResourceClaim events used to trigger a full scan of every unschedulable pod, an O(N²) cost during large 
scale-ups. The DRA plugin now uses a pod informer index to narrow that to just the pods actually affected, 
cutting the requeue path to O(1) and roughly doubling scheduling throughput in early benchmarks. Controlled 
by the `SchedulerPreQueueingHints` feature gate.

## What’s next

DRA continues to mature with every release. Several features currently in alpha and Beta are on track to progress in the 
coming releases, and the community keeps working on DRA's performance, scalability, and reliability. Expect another 
ambitious set of DRA features in Kubernetes 1.38.

## Getting involved

> A good starting point is joining the WG Device Management [Slack channel](https://kubernetes.slack.com/archives/C0409NGC1TK) and [meetings](https://docs.google.com/document/d/1qxI87VqGtgN7EAJlqVfxx86HGKEAc2A3SKru8nJHNkQ/edit?tab=t.0#heading=h.tgg8gganowxq) which happens at US/EU and EU/APAC friendly time slots.

Reviewer 2: nit: I would hyperlink "meetings" to https://www.kubernetes.dev/community/community-groups/wg/device-management/#meetings

Not all enhancement ideas are tracked as issues yet, so come talk to us if you want to help or have some ideas yourself! We have work to do at all levels, from difficult core changes to usability enhancements in `kubectl` which could be picked up by newcomers.

## Acknowledgments

> The following KEP owners managed to add or promote a feature in the 1.37 release (in alphabetic order):

Reviewer 2: (personal note) I am not keen on "managed to"; it can be heard as criticism of any person or group who did not. 

However, it's OK to leave this wording as is.

[Resource Claim Status with possible standardized network interface data](https://www.kubernetes.dev/resources/keps/4817/) adds a `Devices` field to `ResourceClaim.Status`, letting DRA drivers report per-device status, including, for network devices, the interface name, MAC address, and IP addresses. This gives users and controllers visibility into device state that was previously invisible once a device was configured in a Pod, and makes it possible to build things like network services that rely on a device's reported IPs.

[DRA: device taints and tolerations](https://www.kubernetes.dev/resources/keps/5055/) is now stable, DRA drivers can  mark devices as tainted so they're skipped for new Pod scheduling, and cluster admins can apply the same taints cluster-wide via a `DeviceTaintRule`, without reconfiguring drivers. Pods already using a tainted device can be evicted automatically, unless their `ResourceClaim` explicitly tolerates the taint. This mirrors node taints and tolerations, letting operators take a single device offline for maintenance or mark it degraded, without disrupting the rest of the cluster.

> [Standard numaNode device attribute](https://github.com/kubernetes/enhancements/issues/6072) standardizes `resource.kubernetes.io/numaNode` as a shared attribute name, so devices from different drivers can be compared on the same NUMA node instead of each driver inventing its own name for it. It landed directly as stable in 1.37, since it's a naming/registration KEP with no feature gate or in-tree behavior change.

Reviewer 4: ```suggestion
[Standard numaNode device attribute](https://github.com/kubernetes/enhancements/issues/6072) standardizes `resource.kubernetes.io/numaNode` as a shared attribute name, so devices from different drivers can be compared on the same NUMA node instead of each driver inventing its own name for it. It landed directly as stable in v1.37, since it's a naming/registration KEP with no feature gate or in-tree behavior change.
```
Make sure to apply this in all references for the release

## Feature promoted to beta

[ResourceClaim support for workloads](https://www.kubernetes.dev/resources/keps/5729)
graduates to beta behind the `DRAWorkloadResourceClaims` feature gate, which stays disabled by default. 
In a cluster that has the feature enabled, Workloads and PodGroups can reference ResourceClaims directly, so 
a single claim can be shared across an entire group of Pods. This is instead of claims being capped at 256 
Pods through the old per-Pod reservation limit.

The [DRA Device Attributes Downward API](https://www.kubernetes.dev/resources/keps/5304/) is aimed at 
supporting device injection into KubeVirt VMs. Drivers populate a `Metadata` field when preparing a claim, 
and the framework writes it to a JSON file mounted into the container via CDI, letting workloads read a 
device's PCI bus address, MAC address, and other attributes directly instead of requiring custom controllers 
to watch and translate ResourceClaims and ResourceSlices.

> ## Alpha features

Reviewer 4: Make sure all occurrences are using Alpha with capital A

[List types for attributes](https://www.kubernetes.dev/resources/keps/5491) moved into a second alpha in 1.
37, letting a device attribute hold more than one value instead of a single scalar, like a CPU that's 
adjacent to more than one PCIe root. This makes it possible to match or distinguish devices based on 
overlapping or non-overlapping sets of values, while single-value attributes keep working as they do today.

[Node allocatable resource requests](https://www.kubernetes.dev/resources/keps/5517) moved into alpha2. It 
lets the scheduler and kubelet treat DRA-managed CPU, memory, and similar node resources the same way they 
treat ordinary resource requests, so a node doesn't get oversubscribed and users no longer have to duplicate 
the same request in both a ResourceClaim and the pod spec.

[Resource availability visibility](https://www.kubernetes.dev/resources/keps/5677) lets a user see how much 
device capacity is actually left in a pool, not just the total. `kubectl describe resourceslice` and `kubectl 
describe node` are planned to show that directly. It moved into a second alpha in 1.37 to make those numbers 

> accurate for devices that can be shared or split across workloads.

Reviewer 3: https://github.com/kubernetes/enhancements/issues/5677

The description of Resource Availability Visibility here is not accurate. KEP-5677 does not expose live availability data directly through kubectl describe resourceslice or kubectl describe node.
Instead, it uses a one-shot ResourcePoolStatusRequest workflow:

1. Create a ResourcePoolStatusRequest.
2. The controller computes a point-in-time snapshot and writes it to the request’s status.
3. Retrieve the result, typically with kubectl get resourcepoolstatusrequest <name> -o yaml.
4. To refresh the data, delete the request and create a new one.

Therefore, this should not be described as a continuously updated monitoring interface. I’d recommend using the [Resource availability visibility section of the Kubernetes v1.36 DRA update](https://github.com/kubernetes/website/blob/26abefdcfecb7347445180ad2b606097ff409930/content/en/blog/_posts/2026/dra-136-update.md?plain=1#L113-L121) as a reference.

https://github.com/kubernetes/website/blob/26abefdcfecb7347445180ad2b606097ff409930/content/en/blog/_posts/2026/dra-136-update.md?plain=1#L113-L121

> accurate for devices that can be shared or split across workloads.

Reviewer 3: ```suggestion
[Resource availability visibility](https://www.kubernetes.dev/resources/keps/5677) moved to a second alpha in Kubernetes 1.37. Users create a ResourcePoolStatusRequest to get a point-in-time availability snapshot. To refresh it, delete and recreate the request; it is not a continuous monitoring API.
```
Just an example.

Author: Thank you for pointing out , fixed it

tiers based on their available capacity.

[DRA Device Compatibility Groups](https://www.kubernetes.dev/resources/keps/5963) lets drivers tag partitions 
of a device, like MIG vs vGPU profiles on the same GPU, with compatibility groups, so the scheduler rejects 
incompatible combinations up front instead of the driver failing at node preparation time. It's controlled by 
the `DRADeviceCompatibilityGroups` feature gate, disabled by default.

[PreQueueingHint extension point](https://www.kubernetes.dev/resources/keps/6132) is new as alpha in 1.37. 
DRA ResourceClaim events used to trigger a full scan of every unschedulable pod, an O(N²) cost during large 
scale-ups. The DRA plugin now uses a pod informer index to narrow that to just the pods actually affected, 
cutting the requeue path to O(1) and roughly doubling scheduling throughput in early benchmarks. Controlled 
by the `SchedulerPreQueueingHints` feature gate.

## What’s next

DRA continues to mature with every release. Several features currently in alpha and Beta are on track to progress in the 
coming releases, and the community keeps working on DRA's performance, scalability, and reliability. Expect another 

> ambitious set of DRA features in Kubernetes 1.38.

Reviewer 4: I would say to be expected on a future release I wouldn't use 1.38 here

---
layout: blog
title: 'Kubernetes v1.37: DRA Updates'
draft: true
slug: kubernetes-v1-37-dra-updates
author: >
  [Kashish Verma](https://github.com/KashishV999)
---

> Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) keeps pushing past where it started! This release brings DRA Extended Resource support to GA, a milestone the team has been building toward for three straight releases, alongside several more features graduating directly to Stable. A couple of features move to beta and a fresh batch of alpha features rounds out the release.

Reviewer 4: ```suggestion
Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) keeps pushing past where it started! This release brings DRA Extended Resource support to GA, a milestone the team has been building toward for three straight releases, alongside several more features graduating directly to Stable. A couple of features move to Beta and a fresh batch of Alpha features rounds out the release.
```

Author: we use upper camel case only for api Kinds , so for alpha , beta and stable I believe it should be okay

Ref blog : https://kubernetes.io/blog/2025/09/01/kubernetes-v1-34-dra-updates/

> Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) keeps pushing past where it started! This release brings DRA Extended Resource support to GA, a milestone the team has been building toward for three straight releases, alongside several more features graduating directly to stable. A couple of features move to Beta and a fresh batch of alpha features rounds out the release.

Reviewer: ```suggestion
Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) keeps pushing past where it started! This release brings DRA Extended Resource support to GA, a milestone the team has been building toward for three straight releases.  Several more features graduate to beta or GA. A fresh batch of alpha features rounds out the release.
```

I'll dive into what's new for DRA in Kubernetes 1.37!

> ## What's stable in 1.37

Reviewer 4: ```suggestion
## What's Stable in 1.37
```

Author: We should use sentence case for headings , I believe original is okay

[DRA Extended Resource support](https://www.kubernetes.dev/resources/keps/5004) has graduated to GA. This is the mechanism that lets DRA drivers satisfy requests made through the traditional extended resource API, think `example.com/gpu` in a Pod spec, without requiring a separate device plugin alongside the DRA driver. An extended resource name can be set directly on a DeviceClass, and Pods requesting it get matched to a device through DRA with no ResourceClaim needed on the workload's part.

> It's been on a steady path since KEP acceptance in 1.34. Alpha landed in 1.35, beta in 1.36, and now it's stable. For cluster operators, this is what makes DRA adoption gradual. Existing workloads written against extended resources keep working unmodified while the backend allocation logic moves over to DRA.

Reviewer 4: ```suggestion
It's been on a steady path since KEP acceptance in 1.34. Alpha landed in 1.35, Beta in 1.36, and now it's Stable. For cluster operators, this is what makes DRA adoption gradual. Existing workloads written against extended resources keep working unmodified while the backend allocation logic moves over to DRA.
```

Author: upper camel case only for api kinds , I believe not for maturity stage

[Resource Claim Status with possible standardized network interface data](https://www.kubernetes.dev/resources/keps/4817/) adds a `Devices` field to `ResourceClaim.Status`, letting DRA drivers report per-device status, including, for network devices, the interface name, MAC address, and IP addresses. This gives users and controllers visibility into device state that was previously invisible once a device was configured in a Pod, and makes it possible to build things like network services that rely on a device's reported IPs.

> [DRA: device taints and tolerations](https://www.kubernetes.dev/resources/keps/5055/) now Stable, DRA drivers can now mark devices as tainted so they're skipped for new pod scheduling, and cluster admins can apply the same taints cluster-wide via a `DeviceTaintRule`, without reconfiguring drivers. Pods already using a tainted device can be evicted automatically, unless their `ResourceClaim` explicitly tolerates the taint. This mirrors node taints and tolerations, letting operators take a single device offline for maintenance or mark it degraded, without disrupting the rest of the cluster.

Reviewer 4: ```suggestion
[DRA: device taints and tolerations](https://www.kubernetes.dev/resources/keps/5055/) is now Stable. DRA drivers can mark devices as tainted so they're skipped for new Pod scheduling, and cluster admins can apply the same taints cluster-wide via a `DeviceTaintRule`, without reconfiguring drivers. Pods already using a tainted device can be evicted automatically, unless their `ResourceClaim` explicitly tolerates the taint. This mirrors node taints and tolerations, letting operators take a single device offline for maintenance or mark it degraded, without disrupting the rest of the cluster.
```

graduates to beta behind the `DRAWorkloadResourceClaims` feature gate, which stays disabled by default. 
In a cluster that has the feature enabled, Workloads and PodGroups can reference ResourceClaims directly, so 
a single claim can be shared across an entire group of Pods. This is instead of claims being capped at 256 
Pods through the old per-Pod reservation limit.

The [DRA Device Attributes Downward API](https://www.kubernetes.dev/resources/keps/5304/) is aimed at 
supporting device injection into KubeVirt VMs. Drivers populate a `Metadata` field when preparing a claim, 
and the framework writes it to a JSON file mounted into the container via CDI, letting workloads read a 
device's PCI bus address, MAC address, and other attributes directly instead of requiring custom controllers 
to watch and translate ResourceClaims and ResourceSlices.

[DRA Consumable Capacity](https://www.kubernetes.dev/resources/keps/5075/) lets independent resource claims 
draw shares of capacity from the same device instead of requiring exclusive allocation, so a network device's 
bandwidth or a virtual GPU's memory can be split across unrelated workloads. Stable is targeted for 1.38.

> 

Reviewer: We did not promote it this cycle, so it shouldn't be mentioned? Same for the rest of the section. 

The alternative is to rename the section to "features in beta". I prefer the current "feature promoted" because otherwise we end up summarizing the same unchanged features in each feature blog.

Author: good call, you're right

[DRA Partitionable Devices](https://www.kubernetes.dev/resources/keps/4815/) also stays in Beta. It lets a 
device like a GPU be dynamically split into smaller partitions on demand, and extends to multi-host devices 
like interconnected TPUs, so a workload can request a specific slice topology instead of scattered devices.

[Resource Health Status in Pod Status](https://www.kubernetes.dev/resources/keps/4680/) exposes device health 
directly in Pod Status, whether the device came from DRA or the traditional Device Plugin API, so a 
crash-looping Pod using a failed GPU shows up as unhealthy instead of leaving you to guess why it keeps 
restarting.

[DRA Device Binding Conditions](https://www.kubernetes.dev/resources/keps/5007/) lets the scheduler defer 
binding a Pod to a node until an external device is actually ready, useful for fabric-attached GPUs or FPGAs 
that need time to attach or reprogram before use. Instead of binding early and letting the Pod fail or 
crash-loop on a device that isn't there yet, the scheduler waits on a set of conditions the driver defines, 
and reschedules the Pod if preparation fails or times out.

## Alpha features

> [List types for attributes](https://www.kubernetes.dev/resources/keps/5491) moved into a second alpha in 1.
> 37, letting a device attribute hold more than one value instead of a single scalar, like a CPU that's 

Reviewer 4: ```suggestion
[List types for attributes](https://www.kubernetes.dev/resources/keps/5491) moved into a second Alpha in
1.37, letting a device attribute hold more than one value instead of a single scalar, like a CPU that's
```

Author: upper camel case only for api kinds , I believe not for maturity stage

Reviewer 4: https://kubernetes.io/docs/contribute/style/style-guide/#use-start-case-for-enhancement-graduation-phases

adjacent to more than one PCIe root. This makes it possible to match or distinguish devices based on 
overlapping or non-overlapping sets of values, while single-value attributes keep working as they do today.

[Node allocatable resource requests](https://www.kubernetes.dev/resources/keps/5517) moved into alpha2. It 
lets the scheduler and kubelet treat DRA-managed CPU, memory, and similar node resources the same way they 
treat ordinary resource requests, so a node doesn't get oversubscribed and users no longer have to duplicate 
the same request in both a ResourceClaim and the pod spec.

[Resource availability visibility](https://www.kubernetes.dev/resources/keps/5677) lets a user see how much 
device capacity is actually left in a pool, not just the total. `kubectl describe resourceslice` and `kubectl 
describe node` are planned to show that directly. It moved into a second alpha in 1.37 to make those numbers 

> accurate for devices that can be shared or split across workloads.

Reviewer 4: ```suggestion
device capacity is actually left in a pool, not just the total. `kubectl describe resourceslice` and
`kubectl describe node` are planned to show that directly. It moved into a second Alpha in 1.37 to make
those numbers accurate for devices that can be shared or split across workloads.
```

[DRA: Optional Node Operations](https://www.kubernetes.dev/resources/keps/5945) lets a driver skip kubelet's 
prepare and unprepare calls for allocations that don't need any setup on the node. This makes it possible to 
avoid an unnecessary dependency on the driver for allocations where there's genuinely nothing for it to do 
locally.

> [Derived Attributes](https://www.kubernetes.dev/resources/keps/6080) is a new feature that lets you use [CEL](https://kubernetes.io/docs/reference/using-api/cel/) expressions to match up devices based on your own 

Reviewer 4: ```suggestion
[Derived Attributes](https://www.kubernetes.dev/resources/keps/6080) is a new feature, behind the `DRADerivedAttributes` feature gate, that lets you use [CEL](/docs/reference/using-api/cel/) expressions to match up devices based on your own
```

custom rules. Before this, pairing devices from different vendors (like a GPU/TPU and a NIC on the same NUMA 
node) only worked if both drivers used the exact same attribute name. If one used `numa` and the other used 
`numaNode`, the scheduler couldn't pair them together. Now, you can easily bridge these differences yourself 
inside your manifest, meaning you don't have to wait for hardware vendors to agree on standardized attribute 
names. Beyond just fixing naming differences, you can also use CEL to handle more complex scenarios like 
slicing a specific ID out of a long, monolithic topology string, or grouping devices into custom performance 
tiers based on their available capacity.

[DRA Device Compatibility Groups](https://www.kubernetes.dev/resources/keps/5963) lets drivers tag partitions 
of a device, like MIG vs vGPU profiles on the same GPU, with compatibility groups, so the scheduler rejects 
incompatible combinations up front instead of the driver failing at node preparation time. It's controlled by 
the `DRADeviceCompatibilityGroups` feature gate, disabled by default.

> [PreQueueingHint extension point](https://www.kubernetes.dev/resources/keps/6132) is new as alpha in 1.37. 
> DRA ResourceClaim events used to trigger a full scan of every unschedulable pod, an O(N²) cost during large 
> scale-ups. The DRA plugin now uses a pod informer index to narrow that to just the pods actually affected, 
> cutting the requeue path to O(1) and roughly doubling scheduling throughput in early benchmarks. Controlled 
> by the `SchedulerPreQueueingHints` feature gate.

Reviewer 4: ```suggestion
[PreQueueingHint extension point](https://www.kubernetes.dev/resources/keps/6132) is new as Alpha in 1.37.
DRA ResourceClaim events used to trigger a full scan of every unschedulable Pod, an O(N²) cost during large
scale-ups. The DRA plugin now uses a Pod informer index to narrow that to just the Pods actually affected,
cutting the requeue path to O(1). Controlled by the `SchedulerPreQueueingHints` feature gate.
```

## What’s next

DRA continues to mature with every release. Several features currently in Alpha and Beta are on track to progress in the 
coming releases, and the community keeps working on DRA's performance, scalability, and reliability. Expect another 
ambitious set of DRA features in Kubernetes 1.38.

## Getting involved

A good starting point is joining the WG Device Management [Slack channel](https://kubernetes.slack.com/archives/C0409NGC1TK) and [meetings](https://docs.google.com/document/d/1qxI87VqGtgN7EAJlqVfxx86HGKEAc2A3SKru8nJHNkQ/edit?tab=t.0#heading=h.tgg8gganowxq) which happens at US/EU and EU/APAC friendly time slots.

> Not all enhancement ideas are tracked as issues yet, so come talk to us if you want to help or have some ideas yourself! We have work to do at all levels, from difficult core changes to usability enhancements in kubectl which could be picked up by newcomers.

Reviewer 4: ```suggestion
Not all enhancement ideas are tracked as issues yet, so come talk to us if you want to help or have some ideas yourself! We have work to do at all levels, from difficult core changes to usability enhancements in `kubectl` which could be picked up by newcomers.
```

## Acknowledgments

* Patrick Ohly ([pohly](https://github.com/pohly))
* John Belamaric ([johnbelamaric](https://github.com/johnbelamaric))
* Kevin Klues ([klueska](https://github.com/klueska))
* John A. Hull ([johnahull](https://github.com/johnahull))  
* Praveen Krishna ([pravk03](https://github.com/pravk03)) 
* Jiefeng Xu ([jiefeng-xu](https://github.com/jiefeng-xu)) 
* Jon Huhn ([nojnhuh](https://github.com/nojnhuh)) 
* Troy Chiu ([troychiu](https://github.com/troychiu))
* Gaurav Ghildiyal ([gauravkghildiyal](https://github.com/gauravkghildiyal))
* Shingo Omura ([everpeace](https://github.com/everpeace))
* Byonggon Chun([bg-chun](https://github.com/bg-chun))

> * Alay Patel ([alaypatel07](https://github.com/alaypatel07))

Reviewer: The following contributors 
```suggestion

The following KEP owners managed to add or promote a feature in the 1.37 release (in alphabetic order):

* Alay Patel ([alaypatel07](https://github.com/alaypatel07))
* Byonggon Chun([bg-chun](https://github.com/bg-chun))
* Gaurav Ghildiyal ([gauravkghildiyal](https://github.com/gauravkghildiyal))
* Jiefeng Xu ([jiefeng-xu](https://github.com/jiefeng-xu))
* John A. Hull ([johnahull](https://github.com/johnahull))
* Jon Huhn ([nojnhuh](https://github.com/nojnhuh))
* Lionel Jouin ([LionelJouin](https://github.com/LionelJouin))
* Patrick Ohly ([pohly](https://github.com/pohly))
* Praveen Krishna ([pravk03](https://github.com/pravk03))
* Shingo Omura ([everpeace](https://github.com/everpeace))
* Troy Chiu ([troychiu](https://github.com/troychiu))
```

Reviewer: Note that this includes Praveen's alpha1 -> alpha2 transition, which is a kind of "promotion".

> And a huge thanks to everyone else who helped shape this release, in ways big and small. Given enough eyeballs, all bugs are shallow and this release had plenty of them, watching closely and caring enough to make things better. DRA got better this cycle because of all of you.

Reviewer: ```suggestion
This would not have been possible without the help of the reviewers and approvers.
So a huge thanks to everyone else who helped shape this release, in ways big and small. Given enough eyeballs, all bugs are shallow and this release had plenty of them, watching closely and caring enough to make things better. DRA got better this cycle because of all of you.
```

> And a huge thanks to everyone else who helped shape this release, in ways big and small. Given enough eyeballs, all bugs are shallow and this release had plenty of them, watching closely and caring enough to make things better. DRA got better this cycle because of all of you.

Reviewer 4: ```suggestion
And a huge thanks to everyone else who helped shape this release, in ways big and small. Given enough eyeballs, all bugs are shallow — and this release had plenty of eyeballs on it, watching closely and caring enough to make things better. DRA got better this cycle because of all of you.
```

> And a huge thanks to everyone else who helped shape this release, in ways big and small. Given enough eyeballs, all bugs are shallow and this release had plenty of them, watching closely and caring enough to make things better. DRA got better this cycle because of all of you.

Author: I believe this is original reads okay

---
layout: blog
title: 'Kubernetes v1.37: DRA Updates'
draft: true
slug: kubernetes-v1-37-dra-updates
author: >
  Kashish Verma
---

> Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)] (/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) 

Reviewer 2: By the time this publishes, that link may be to a redirect (the DRA page is moving URL).

I think we leave it as is but make a note to revisit after the v1.37 release has happened.

Author: Okay : Noted

> Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)] (/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) 

Reviewer 5: ```suggestion
Kubernetes 1.37 is here and [Dynamic Resource Allocation (DRA)](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) 
```

keeps pushing past where it started! This release brings DRA Extended Resource support to GA, a milestone the team has been 
building toward for three straight releases, alongside a new standard device attribute that lands directly as Stable. 
One feature graduates to beta and most of the alpha work continues into a second round, alongside new additions.

> Let's dive into what's new for DRA in Kubernetes 1.37!

Reviewer 2: (nit)
```suggestion
I'll dive into what's new for DRA in Kubernetes 1.37!
```

## Extended Resource allocation graduates to GA

> [DRA Extended Resource support](https://github.com/kubernetes/enhancements/issues/5004) has graduated to GA. This is the mechanism that lets DRA drivers satisfy requests made through the traditional extended resource API, think `example.com/gpu`  in a Pod spec, without requiring a separate device plugin alongside the DRA driver. An extended resource name can be set directly on a DeviceClass, and Pods requesting it get matched to a device through DRA with no ResourceClaim needed on the workload's part.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5004 here.

> It's been on a steady path since KEP acceptance in 1.34. Alpha landed in 1.35, beta in 1.36, and now it's stable. For cluster operators, this is what makes DRA adoption gradual. Existing workloads written against extended resources keep working unmodified while the backend allocation logic moves over to DRA.

Reviewer: https://github.com/kubernetes/enhancements/issues/4817 and https://github.com/kubernetes/enhancements/issues/5055 also graduated to GA, see https://docs.google.com/presentation/d/1COH8dG8qZ9jEMXzQVdco74XmroZQTMYcUgwAES8FYWU/edit?slide=id.g376578994e9_0_0&pli=1&resourcekey=0-RCm3FFt8pxiK-y-_YHUkOg#slide=id.g376578994e9_0_0 lists

Reviewer 3: https://github.com/kubernetes/enhancements/issues/5304 is also missing, which was promoted to beta in v1.37.

Author: Added all the stable keps' and https://github.com/kubernetes/enhancements/issues/5304 , 
now there are these remaining :  but as far as I know the release-comms team communicated with the leads of all these keps' to opt-in for a combined feature-blog but these keps' never opt-in , so are we still planning to include them : if so we need to communicate to release-comms team bout the total number of opt-ins
```
## Remaining
5963 - alpha
6132 - alpha
5075 : beta
4815: beta
4680: beta
5007 : beta
```

Author: I have them ready , just need confirmation so I will update and add them

Reviewer: Writing a "DRA feature blog" that leaves our some work just because the KEP owners failed to respond individually for their KEP feels wrong. I was under the impression that we had communicated our intend to write a combined blog entry for all DRA-related KEPs.

Author: Got it , added !

Reviewer 3: For the remaining items, I think we should still mention any significant changes, even if only in a single sentence.

For example, [KEP-5075](https://github.com/kubernetes/enhancements/issues/5075) added `DRAFractionalCapacityRange` in v1.37, even though the KEP was not tracked as part of the v1.37 release team.

cc @sunya-ch

Reviewer 6: @pacoxu @KashishV999 

I’ve dropped a brief mention here about `DRAFractionalCapacityRange `. Please feel free to use and modify. 

> [DRA Consumable Capacity](https://www.kubernetes.dev/resources/keps/5075/) now supports fractional values in CapacityRequestPolicyRange, enabling more precise capacity requests and allocation for devices with fractional resources. This improves flexibility for workloads that require fine-grained resource allocation. The enhancement is gated by the `DRAFractionalCapacityRange` feature gate, which is in beta in 1.37.

Please let me know if you need further input from me.

PR source: https://github.com/kubernetes/kubernetes/pull/140161

> ## A New Stable Feature

Reviewer 2: nit: should be sentence case

Maybe "Standard attribute for NUMA node names"?

> [Standard numaNode device attribute](https://github.com/kubernetes/enhancements/issues/6072) standardizes `resource.kubernetes.io/numaNode` as a shared attribute name, so devices from different drivers can be compared on the same NUMA node instead of each driver inventing its own name for it. It lands directly as Stable in 1.37, since it's a naming/registration KEP with no feature gate or in-tree behavior change.

Reviewer 2: nit: "landed" (not "lands")

> [Standard numaNode device attribute](https://github.com/kubernetes/enhancements/issues/6072) standardizes `resource.kubernetes.io/numaNode` as a shared attribute name, so devices from different drivers can be compared on the same NUMA node instead of each driver inventing its own name for it. It lands directly as Stable in 1.37, since it's a naming/registration KEP with no feature gate or in-tree behavior change.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/6072 here.

> ## Feature promoted to Beta

Reviewer 2: This heading looks wrong.

Reviewer: ```suggestion
## Features promoted to beta
```

Is this better? We had more than one (plural) and I don't see a reason to write "beta" with capital letter.

> [ResourceClaim support for workloads](https://github.com/kubernetes/enhancements/issues/5729)

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5729 here.

> graduates to beta behind the `DRAWorkloadResourceClaims` feature gate, which stays disabled by default. Workloads and PodGroups can reference ResourceClaims directly, so a single claim can be shared across an entire group of Pods instead of being capped at 256 Pods through the old per-Pod reservation limit.

Reviewer 2: ```suggestion
graduates to beta behind the `DRAWorkloadResourceClaims` feature gate, which stays disabled by default.
In a cluster that has the feature enabled, Workloads and PodGroups can reference ResourceClaims directly,
so a single claim can be shared across an entire group of Pods.
This is instead of claims being capped at 256 Pods through the old per-Pod reservation limit.
```

## Alpha features

> [List types for attributes](https://github.com/kubernetes/enhancements/issues/5491) moved into a second alpha in 1.37, letting a device attribute hold more than one value instead of a single scalar, like a CPU that's adjacent to more than one PCIe root. This makes it possible to match or distinguish devices based on overlapping or non-overlapping sets of values, while single-value attributes keep working as they do today.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5491 here.

> [List types for attributes](https://github.com/kubernetes/enhancements/issues/5491) moved into a second alpha in 1.37, letting a device attribute hold more than one value instead of a single scalar, like a CPU that's adjacent to more than one PCIe root. This makes it possible to match or distinguish devices based on overlapping or non-overlapping sets of values, while single-value attributes keep working as they do today.

Reviewer 7: Thanks for drafting this. The contents of this paragraph is LGTM.

> [Node allocatable resource requests](https://github.com/kubernetes/enhancements/issues/5517) moved into alpha2. It lets the scheduler and kubelet treat DRA-managed CPU, memory, and similar node resources the same way they treat ordinary resource requests, so a node doesn't get oversubscribed and users no longer have to duplicate the same request in both a ResourceClaim and the pod spec.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5517 here.

> [Resource availability visibility](https://github.com/kubernetes/enhancements/issues/5677) lets a user see how much device capacity is actually left in a pool, not just the total. `kubectl describe resourceslice` and `kubectl describe node` are planned to show that directly. It moved into a second alpha in 1.37 to make those numbers accurate for devices that can be shared or split across workloads.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5677 here.

> [DRA: Optional Node Preparation](https://github.com/kubernetes/enhancements/issues/5945) lets a driver skip kubelet's prepare and unprepare calls for allocations that don't need any setup on the node. This makes it possible to avoid an unnecessary dependency on the driver for allocations where there's genuinely nothing for it to do locally.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5945 here.

> [DRA: Optional Node Preparation](https://github.com/kubernetes/enhancements/issues/5945) lets a driver skip kubelet's prepare and unprepare calls for allocations that don't need any setup on the node. This makes it possible to avoid an unnecessary dependency on the driver for allocations where there's genuinely nothing for it to do locally.

Reviewer 5: ```suggestion
[DRA: Optional Node Operations](https://github.com/kubernetes/enhancements/issues/5945) lets a driver skip kubelet's prepare and unprepare calls for allocations that don't need any setup on the node. This makes it possible to avoid an unnecessary dependency on the driver for allocations where there's genuinely nothing for it to do locally.
```
We used Optional Node Operations in the codebase. I just updated the issue https://github.com/kubernetes/enhancements/issues/5945 as well.

Reviewer 2: Ideally, update the `kep.yaml` in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/5945-dra-optional-node-preparation too

> [Derived Attributes](https://github.com/kubernetes/enhancements/issues/6080) is a new feature that lets you use CEL expressions to match up devices based on your own custom rules. Before this, pairing devices from different vendors (like a GPU/TPU and a NIC on the same NUMA node) only worked if both drivers used the exact same attribute name. If one used `numa` and the other used `numaNode`, the scheduler couldn't pair them together. Now, you can easily bridge these differences yourself inside your manifest, meaning you don't have to wait for hardware vendors to agree on standardized attribute names. Beyond just fixing naming differences, you can also use CEL to handle more complex scenarios like slicing a specific ID out of a long, monolithic topology string, or grouping devices into custom performance tiers based on their available capacity.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/6080 here.

Also consider making "CEL" a hyperlink.

> [DRA Device Compatibility Groups](https://github.com/kubernetes/enhancements/issues/5963) is new as alpha in 1.37. Partitionable devices can represent multiple logical devices that consume capacity from the same counter set, but counter accounting alone cannot express that some device modes are mutually incompatible. Drivers can now attach opaque compatibility-group names to each device’s counter-set consumption, and devices consuming from the same counter-set may be co-allocated only when all of them share at least one common group. Devices without groups can only be co-allocated with other ungrouped devices. This lets Kubernetes reject incompatible combinations during scheduling instead of discovering the conflict later during node preparation. The feature is controlled by the `DRADeviceCompatibilityGroups` feature gate, which is disabled by default.

Reviewer 2: Could hyperlink to https://www.kubernetes.dev/resources/keps/5963 here.

> ## What’s next

Reviewer 2: Maybe omit this; alternatively, make the remaining headings be at level 3.

Author: Added content for this heading

## Getting involved

A good starting point is joining the WG Device Management [Slack channel](https://kubernetes.slack.com/archives/C0409NGC1TK) and [meetings](https://docs.google.com/document/d/1qxI87VqGtgN7EAJlqVfxx86HGKEAc2A3SKru8nJHNkQ/edit?tab=t.0#heading=h.tgg8gganowxq) which happens at US/EU and EU/APAC friendly time slots.

Not all enhancement ideas are tracked as issues yet, so come talk to us if you want to help or have some ideas yourself! We have work to do at all levels, from difficult core changes to usability enhancements in kubectl which could be picked up by newcomers.

## Acknowledgments

> A huge thanks to everyone who has contributed:

Reviewer: How is this list sorted?

I'm missing in the list, despite being involved in almost every KEP as reviewer and often also as co-contributor...

Reviewer: Or asking more generally: whom do we want to list here? A lot of people have helped with reviews. Should we list all?

Author: Sorry about that, added it now !. I've also added a note in the PR description so contributors can flag if they're missing
On who to include: I'd say let's stick to leads and core reviewers for now, otherwise the list is going to get really long. open to other ideas though :)

Author: Or in other case : just include the new contributors for this cycle to encourage  more contributions

Reviewer: I'm not sure why you missed adding my name. I'm a WG lead and do most of the day-to-day business. Is my work really so unknown? :thinking: 

But this perhaps illustrates the problem: it's difficult to come up with an objective way to identify who "deserves" to be listed.

I'm leaning towards defining a easily verifiable criteria and than listing those who match. I'm leaning towards "owner of a KEP which graduated to the next level in 1.37". We can then add a sentence saying something like "a huge thanks to all reviewers and approvers who made this possible - they are too numerous to list here".

Reviewer: cc @johnbelamaric

Author: at the end of acknowledgment section I added a line to thank everyone:
"And a huge thanks to everyone else who helped shape this release, in ways big and small.... DRA got better this cycle because of all of you."

* John Belamaric ([johnbelamaric](https://github.com/johnbelamaric))
* Kevin Klues ([klueska](https://github.com/klueska))
* John A. Hull ([johnahull](https://github.com/johnahull))  
* Praveen Krishna ([pravk03](https://github.com/pravk03)) 
* Jiefeng Xu ([jiefeng-xu](https://github.com/jiefeng-xu)) 
* Jon Huhn ([nojnhuh](https://github.com/nojnhuh)) 

> * Troy Chiu ([troychiu])(https://github.com/troychiu)

Reviewer 5: ```suggestion
* Troy Chiu ([troychiu](https://github.com/troychiu))
```

* Gaurav Ghildiyal ([gauravkghildiyal](https://github.com/gauravkghildiyal))

> 

Reviewer 7: Although my contribution is not so large in this release, is it ok to list my name here? I am an owner of KEP-5491 List Type Attribute.

```suggestion
* Shingo Omura ([everpeace](https://github.com/everpeace))
```

Reviewer 8: Yes, of course! Thanks for adding that in as a suggestion. @KashishV999, please have a look :)

Author: added , thank you
